core/src/main/java/org/infinispan/remoting/inboundhandler/BaseBlockingRunnable.java

```diff
@@ -28,10 +28,18 @@ protected BaseBlockingRunnable(BasePerCacheInboundInvocationHandler handler, Cac
 
    @Override
    public void run() {
+      InvocationStatus invocationStatus = null;
       try {
-         response = beforeInvoke();
-         if (response == null) {
-            response = handler.invokePerform(command);
+         invocationStatus = beforeInvoke();
+         switch (invocationStatus) {
+            case CONTINUE:
+               response = handler.invokePerform(command);
+               break;
+            case CACHE_NOT_FOUND:
+               response = CacheNotFoundResponse.INSTANCE;
+               break;
+            case DEFERRED:
+               return;
```

Comment:
@danberindei How does the per-sender ordering keep us away from the problem? If it was causal ordering, I could see that, but per-sender?

I don't see a reason to use synchronous API with `DIST_ASYNC` when the asynchronous API can do the trick efficiently (without context switch and concurrency limitations through delegating the work to limited pool). On the other hand, I can see a lot of space for misuse, when coder's mind set considers only sync mode and administrator turns on `DIST_ASYNC` because it's more performant and 'seems to work'.
I know there's additional perf difference by not requiring the response, but I would rather see that as a flag. But let's keep this for infinispan-dev.